### PR TITLE
feat(sbom): add SHA-256 hash support to CycloneDX SBOM component model

### PR DESCRIPTION
**Pull request by a-oren**
## Summary

- Add optional `hashes` parameter to `getComponent()` and `addDependency()` in the CycloneDX SBOM model
- Update `Sbom` wrapper, `base_pyproject` graph structure, `python_controller` `DependencyEntry`, and `python_pip` SBOM creation paths to forward hash data
- Hashes are omitted when not provided, maintaining full backward compatibility
- Add 6 unit tests covering hashes present, absent, empty array, serialization, and source vs target isolation

Implements [TC-4330](https://redhat.atlassian.net/browse/TC-4330)

## Test plan

- [x] 6 new unit tests for hash support pass
- [x] All existing tests pass unchanged (0 regressions)
- [x] ESLint clean on all modified files

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[TC-4330]: https://redhat.atlassian.net/browse/TC-4330?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Add optional hash support to CycloneDX SBOM components and propagate hash data through Python dependency providers while keeping existing behavior unchanged when hashes are absent.

New Features:
- Allow CycloneDX SBOM components to include optional hashes metadata on components and dependencies.

Enhancements:
- Extend SBOM and Python provider models to carry hash information through graph and dependency structures.
- Update SBOM construction logic so hash data is attached or updated on existing components without creating duplicates.

Tests:
- Add unit tests validating hash propagation, omission when absent or empty, JSON serialization, and correct handling when components are first created without hashes.

### CODE REVIEW

**Comment by a-oren** [comment]
@sourcery-ai review

File: src/providers/python_pip.js [hunk 1]
```diff
@@ -84,7 +84,7 @@ async function provideComponent(manifest, opts = {}) {
  */
 function addAllDependencies(source, dep, sbom) {
 	let targetPurl = toPurl(dep["name"], dep["version"])
-	sbom.addDependency(source, targetPurl)
+	sbom.addDependency(source, targetPurl, undefined, dep["hashes"])
```
Comment (ruromero) [inline src/providers/python_pip.js]: Can you create a `const NO_SCOPE=undefined` ? It will make it more evident what is the `undefined` for.
later on we can refactor into an options object

**Review by ruromero — CHANGES_REQUESTED** [review]
Looks good, I'm missing the updated expected python resulting sbom (stack/component) for pip

File: src/cyclone_dx_sbom.js [hunk 2]
```diff
@@ -109,17 +115,20 @@ export default class CycloneDxSbom {
...
+	addDependency(sourceRef, targetRef, scope, targetHashes) {
 		const sourcePurl = sourceRef.toString();
 		const targetPurl = targetRef.toString();
 
 		// Ensure both components exist in the components list
 		[sourceRef, targetRef].forEach((ref, index) => {
 			const purl = index === 0 ? sourcePurl : targetPurl;
 			if (this.getComponentIndex(purl) < 0) {
-				this.components.push(getComponent(ref, "library", scope));
+				const hashes = index === 1 ? targetHashes : undefined;
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/cyclone_dx_sbom.js:131]: [sdlc-workflow/verify-pr] Classified as **code change request** — already addressed in commit d726890. No sub-task created (fix already committed).

File: src/providers/python_pip.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ import { getParser, getIgnoreQuery, getPinnedVersionQuery } from './requirements
 
 export default { isSupported, validateLockFile, provideComponent, provideStack, readLicenseFromManifest }
 
-/** @typedef {{name: string, version: string, dependencies: DependencyEntry[]}} DependencyEntry */
+/** @typedef {{name: string, version: string, dependencies: DependencyEntry[], hashes?: Array<{alg: string, content: string}>}} DependencyEntry */
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/providers/python_pip.js:19]: [sdlc-workflow/verify-pr] Classified as **suggestion** — already consolidated in commit d726890. No sub-task created.

File: test/cyclone_dx_sbom_hashes.test.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	/** Verifies that addDependency with hashes produces a component containing the hashes array. */
+	test('getComponent with hashes includes hashes array in component', () => {
+		// Given an SBOM with a root and a dependency with hashes
+		const sbom = new CycloneDxSbom()
+		const root = new PackageURL('pypi', undefined, 'my-app', '1.0.0', undefined, undefined)
+		const purl = new PackageURL('pypi', undefined, 'requests', '2.33.1', undefined, undefined)
+		sbom.addRoot(root)
+
+		// When adding a dependency with hashes
+		sbom.addDependency(root, purl, undefined, sampleHashes)
```
[the remark this answers is not in the document]
Comment (a-oren) [reply test/cyclone_dx_sbom_hashes.test.js:19]: [sdlc-workflow/verify-pr] Classified as **code change request** — already addressed in commit d726890 with two new tests covering both discovery orderings. No sub-task created.

File: src/cyclone_dx_sbom.js [hunk 5]
```diff
@@ -54,6 +55,11 @@ function getComponent(component, type, scope, licenses) {
 		});
 	}
 
+	// Add hashes if provided (CycloneDX 1.4 format).
+	if (hashes && hashes.length > 0) {
+		componentObject.hashes = hashes;
+	}
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/cyclone_dx_sbom.js:61]: [sdlc-workflow/verify-pr] Classified as **suggestion** — no established codebase convention for defensive copying; existing `licenses` handling follows the same by-reference pattern. No sub-task created.

File: src/providers/base_pyproject.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ const IGNORE_MARKERS = ['exhortignore', 'trustify-da-ignore']
 const DEFAULT_ROOT_NAME = 'default-pip-root'
 const DEFAULT_ROOT_VERSION = '0.0.0'
 
-/** @typedef {{name: string, version: string, children: string[]}} GraphEntry */
+/** @typedef {{name: string, version: string, children: string[], hashes?: Array<{alg: string, content: string}>}} GraphEntry */
 /** @typedef {{name: string, version: string, dependencies: DepTreeEntry[]}} DepTreeEntry */
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/providers/base_pyproject.js:21]: [sdlc-workflow/verify-pr] Classified as **suggestion** — no central types file or shared typedef convention in this codebase. No sub-task created.

File: src/providers/python_pip.js (hunk 1 above)
Comment (a-oren) [reply src/providers/python_pip.js]: [sdlc-workflow/verify-pr] Classified as **code change request** — reviewer requests a `const NO_SCOPE = undefined` for readability. Noted for implementation.

**Comment by a-oren** [comment]
## Verification Report — TC-4330

### Review Feedback
| # | Thread | Author | Classification | Status |
|---|--------|--------|---------------|--------|
| 1 | Components never get hashes updated | sourcery-ai | code change request | ✅ Addressed (d726890) |
| 2 | Duplicate DependencyEntry typedef | sourcery-ai | suggestion | ✅ Addressed (d726890) |
| 3 | Missing test for existing component | sourcery-ai | code change request | ✅ Addressed (d726890) |
| 4 | Clone hashes array (shallow copy) | sourcery-ai | suggestion | Skipped — no codebase convention |
| 5 | Shared HashEntry typedef | sourcery-ai | suggestion | Skipped — no central types file |
| 6 | `const NO_SCOPE = undefined` | ruromero | code change request | Pending implementation |

**Ruben's review body** ("missing updated expected python resulting sbom"): This is a plumbing task — no provider populates hashes yet, so golden files remain unchanged. The existing pip golden file tests pass identically to clean main.

### Deterministic Checks

| Check | Result |
|-------|--------|
| Scope containment | ✅ PASS — 6 files changed, all within src/ and test/ |
| Diff size | ✅ PASS — +183 / -17 lines (small) |
| Commit traceability | ✅ PASS — 2 commits, both reference TC-4330 via branch |
| Sensitive patterns | ✅ PASS — no secrets, credentials, or .env files |
| CI status | ✅ PASS — all checks green (lint+test Node 22/24, PR title, commit messages) |

### Acceptance Criteria

| # | Criterion | Result |
|---|-----------|--------|
| 1 | `getComponent()` accepts optional hashes parameter | ✅ PASS |
| 2 | `addDependency()` accepts and forwards hash data | ✅ PASS |
| 3 | Graph data structure in base_pyproject.js supports optional hashes | ✅ PASS |
| 4 | DependencyEntry in python_controller.js supports optional hashes | ✅ PASS |
| 5 | Without hashes, components generated without hashes field | ✅ PASS |
| 6 | Existing tests continue to pass unchanged | ✅ PASS |

### Test Requirements

| # | Requirement | Result |
|---|------------|--------|
| 1 | Unit test: getComponent with hashes | ✅ PASS |
| 2 | Unit test: getComponent without hashes | ✅ PASS |
| 3 | Unit test: addDependency forwards hashes | ✅ PASS |
| 4 | All existing SBOM golden file tests pass | ✅ PASS |

### Overall Result: **WARN**
All acceptance criteria pass. One open code change request from @ruromero (`NO_SCOPE` constant) pending implementation. Bot suggestions (shallow copy, shared typedef) classified as non-actionable per codebase conventions.

---
_This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.5.11._

**Review by ruromero — APPROVED** [review]
lgtm
